staticDHCPd/staticdhcpdlib/databases/_caching.py

```diff
@@ -163,15 +163,67 @@ def _cacheMAC(self, mac, definition, chained):
          definition.domain_name, definition.domain_name_servers, definition.ntp_servers,
          definition.lease_time
         )
-        
+
+class MemcachedCache(_DatabaseCache):
+    """
+    A memory database cache using memcache.
+    """
+    _mac_cache = None #: A dictionary of cached MACs
+    _subnet_cache = None #: A dictionary of cached subnet/serial data
+
+    def __init__(self, name, memcached_server, memcached_age_time, chained_cache=None):
+        """
+        Initialises the cache.
+
+        :param basestring name: The name of the cache.
+        :param basestring memcached_server: Address and port to connect to the memcached server.
+        :param basestring memcached_age_time: number of seconds to store items in memcache.
+        :param :class:`_DatabaseCache <_DatabaseCache>` chained_cache: The next
+            node in the chain; None if this is the end.
+        """
+        _DatabaseCache.__init__(self, name, chained_cache=chained_cache)
+        import memcache
+
+        self.mc_client = memcache.Client([memcached_server])
+        self.memcached_age_time = memcached_age_time
+        _logger.debug("Memcached database-cache initialised")
+
+    def _reinitialise(self):
+        self.mc_client.flush_all()
+
+    def _lookupMAC(self, mac):
+        data = self.mc_client.get(str(mac))
+        if data:
+            (ip, hostname, extra, subnet_id) = data
+            subnet_str = "%s-%s" % subnet_id
+            details = self.mc_client.get(subnet_str)
+            return Definition(
```

Comment:
We'll probably want to check to make sure that we actually get values for 'details' (could use a better name) here, since it's independent and with memcache, things can expire. So it's *possible* (but probably not lilkely) for the mac to be there but not the memcache and issue like that would be hard to track down, so it would be better to check for details before accessing it.